demos/nwqn_data_pull/retrieve_nwqn_streamflow.py

```diff
@@ -0,0 +1,134 @@
+
+import lithops
+import os
+import numpy as np
+import pandas as pd
+
+
+from dataretrieval import nwis
+from random import randint
+from time import sleep
+
+from retrieve_nwqn_samples import find_neighboring_sites, BAD_NLDI_SITES
+
+DESTINATION_BUCKET = os.environ.get('DESTINATION_BUCKET')
+START_DATE = "1991-01-01"
+END_DATE = "2023-12-31"
+
+def map_retrieval(site):
+    """Map function to pull data from NWIS and WQP"""
+    print(f"Retrieving samples from site {site}")
+
+    if site in BAD_NLDI_SITES:
+        site_list = [site]
+    # else query slowly
+    else:
+        sleep(randint(0, 5))
+        site_list = find_neighboring_sites(site)
+
+    df, _ = nwis.get_dv(
+        sites=site_list,
+        start=START_DATE,
+        end=END_DATE,
+        parameterCd="00060",
+    )
+
+    # by default, site_no is not in the index if a single site is queried
+    if "site_no" in df.columns:
+        index_name = df.index.names[0]
+        df.set_index(["site_no", df.index], inplace=True)
+        df.index.names = ["site_no", index_name]
+
+    print(len(df), "records retrieved")
+    # process the results
+    if not df.empty:
+        # fill missing codes to enable string operations
+        df["00060_Mean_cd"] = df["00060_Mean_cd"].fillna("M")
+        df = df[df["00060_Mean_cd"].str.contains("A")]
+        df['00060_Mean'] = df['00060_Mean'].replace(-999999, np.nan)
+
+        site_info, _ = nwis.get_info(sites=site_list)
+        # USACE sites may have same site_no, which creates index conflicts later
+        site_info = site_info[site_info["agency_cd"] == "USGS"]  # keep only USGS sites
+        site_info = site_info.set_index("site_no")
+
+        main_site = site_info.loc[site]
+        main_site_drainage_area = main_site["drain_area_va"]
+
+        # compute fraction of drainage area
+        site_info = site_info[["drain_area_va"]].copy()
+        site_info["drain_fraction"] = site_info["drain_area_va"] / main_site_drainage_area
+        site_info["fraction_diff"] = np.abs(1 - site_info["drain_fraction"])
+
+        # apply drainage area fraction
+        df = pd.merge(df, site_info, left_index=True, right_index=True)
+        df["00060_Mean"] *= site_info.loc[df.index.get_level_values("site_no"), "drain_fraction"].values
+
+        # order sites for filling in missing data
+        fill_order = site_info.sort_values("fraction_diff", ascending=True)
+        fill_order = fill_order.index.values
+
+        flow_sites = df.index.get_level_values("site_no").values
+        fill_order = set(fill_order).intersection(flow_sites)
+
+        output = pd.DataFrame()
+
+        # loop through sites, updating the best data first
+        for fill_site in fill_order:
+            fill_data = df.loc[fill_site]
+            output = update_dataframe(output, fill_data)
+
+        output = output.drop(columns=["drain_area_va", "drain_fraction", "fraction_diff"])
+        output["site_no"] = site
+
```

Comment:
I tried running this function line-by-line for site '01170095', which had 3 nearby sites: ['01170070', '01170095', '01170100', '01170103']. It appears that the only site with data was '01170070', so that site was wholly used to fill in for site '01170095'. Is this the desired outcome?

Reply:
yes